File: simplypdf/src/main/java/com/wwdablu/soumya/simplypdf/SimplyPdf.java
package com.wwdablu.soumya.simplypdf;

import android.content.Context;
import android.print.PrintAttributes;

import androidx.annotation.NonNull;

import com.wwdablu.soumya.simplypdf.jsonengine.SimplyJson;

import java.io.File;

import io.reactivex.Observable;

public final class SimplyPdf {

    private final SimplyPdfDocument document;

    private SimplyPdf(@NonNull Context context, @NonNull File outputPdf) {
        document = new SimplyPdfDocument(context, outputPdf);
    }

    public static SimplyPdf with(@NonNull Context context, @NonNull File outputPdf) {
        return new SimplyPdf(context, outputPdf);
    }

    public static Observable<Boolean> use(@NonNull Context context,
                                           @NonNull SimplyPdf simplyPdf,
                                           @NonNull String payload) {

        return new SimplyJson(context, payload).generateWith(simplyPdf.document);
    }

    public SimplyPdf colorMode(DocumentInfo.ColorMode colorMode) {
        document.getDocumentInfo().setColorMode(colorMode);
        return this;
    }

    public SimplyPdf margin(DocumentInfo.Margins margins) {
        document.getDocumentInfo().setMargins(margins);
        return this;
    }

    public SimplyPdf paperSize(PrintAttributes.MediaSize paperSize) {
        document.getDocumentInfo().setPaperSize(paperSize);
        return this;
    }

    public SimplyPdf paperOrientation(DocumentInfo.Orientation orientation) {
        document.getDocumentInfo().setOrientation(orientation);
        return this;
    }

    public SimplyPdfDocument build() {
        document.build();
        return document;
    }
}

File: simplypdf/src/main/java/com/wwdablu/soumya/simplypdf/SimplyPdf.kt
package com.wwdablu.soumya.simplypdf

import android.content.Context
import android.print.PrintAttributes
import com.wwdablu.soumya.simplypdf.document.DocumentInfo
import com.wwdablu.soumya.simplypdf.document.DocumentInfo.ColorMode
import com.wwdablu.soumya.simplypdf.document.Margin
import com.wwdablu.soumya.simplypdf.document.PageModifier
import com.wwdablu.soumya.simplypdf.jsonengine.SimplyJson
import com.wwdablu.soumya.simplypdf.jsonengine.base.ComposerConverter
import com.wwdablu.soumya.simplypdf.jsonengine.base.PageConverter
import java.io.File

class SimplyPdf private constructor(private val context: Context, outputPdf: File) {

    private val document: SimplyPdfDocument = SimplyPdfDocument(context, outputPdf)

    fun colorMode(colorMode: ColorMode): SimplyPdf {
        document.documentInfo.colorMode = colorMode
        return this
    }

    fun margin(margins: Margin): SimplyPdf {
        document.documentInfo.margins = margins
        return this
    }

    fun paperSize(paperSize: PrintAttributes.MediaSize): SimplyPdf {
        document.documentInfo.paperSize = paperSize
        return this
    }

    fun paperOrientation(orientation: DocumentInfo.Orientation): SimplyPdf {
        document.documentInfo.orientation = orientation
        return this
    }

    fun pageModifier(pageModifier: PageModifier) : SimplyPdf {
        document.pageModifiers.add(pageModifier)
        return this
    }

    fun build(): SimplyPdfDocument {
        document.build(context)
        return document
    }

    companion object {
        @kotlin.jvm.JvmStatic
        fun with(context: Context, outputPdf: File): SimplyPdf {
            return SimplyPdf(context, outputPdf)
        }

        @kotlin.jvm.JvmStatic
        suspend fun usingJson(
            context: Context,
            simplyPdfDocument: SimplyPdfDocument,
            payload: String,
            pageConverters: List<PageConverter>? = null,
            composerConverters: List<ComposerConverter>? = null
        ) {
            SimplyJson(context, payload).generateWith(simplyPdfDocument,
                pageConverters, composerConverters)
        }

        suspend fun usingJson(
            context: Context,
            outputPdf: File,
            payload: String,
            pageConverters: List<PageConverter>? = null,
            composerConverters: List<ComposerConverter>? = null
        ) {
            SimplyJson(context, payload).generateWith(SimplyPdf(context, outputPdf),
                pageConverters, composerConverters)
        }
    }
}